frame/balances/src/lib.rs

```diff
@@ -1656,23 +1646,40 @@ where
 
 	type MaxLocks = T::MaxLocks;
 
+	// Checks if `set_lock` or `extend_lock` would fail. Specifically if adding a lock with `id`
+	// would push us over the `MaxLocks` limit.
+	fn can_add_lock(id: LockIdentifier, who: &T::AccountId) -> bool {
+		let locks = Self::locks(who);
+		let length = locks.len();
+		// If length < MaxLocks, we can always add one more new lock
+		if (length as u32) < T::MaxLocks::get() { return true }
+		// If `length == T::MaxLocks::get()`, and we check if this id is already used, which
+		// means we can "add" this lock since it would update an existing id.
+		if locks.into_iter().any(|l| l.id == id) { return true }
+		return false
+	}
+
 	// Set a lock on the balance of `who`.
 	// Is a no-op if lock amount is zero or `reasons` `is_none()`.
 	fn set_lock(
 		id: LockIdentifier,
 		who: &T::AccountId,
 		amount: T::Balance,
 		reasons: WithdrawReasons,
-	) {
-		if amount.is_zero() || reasons.is_empty() { return }
+	) -> DispatchResult {
+		if amount.is_zero() || reasons.is_empty() { return Ok(()) }
 		let mut new_lock = Some(BalanceLock { id, amount, reasons: reasons.into() });
 		let mut locks = Self::locks(who).into_iter()
 			.filter_map(|l| if l.id == id { new_lock.take() } else { Some(l) })
 			.collect::<Vec<_>>();
 		if let Some(lock) = new_lock {
 			locks.push(lock)
 		}
-		Self::update_locks(who, &locks[..]);
+		let bounded_locks = BoundedVec::<_, T::MaxLocks>::try_from(
+			locks.to_vec(),
+		).map_err(|_| Error::<T, I>::TooManyLocks)?;
```

Comment:
There are two different reallocations of the entire vector here. We can avoid those:

```rust
let mut locks = Self::locks(who).try_mutate(|locks| {
  for existing in locks.iter_mut() {
    if existing.id == id {
      *existing = new_lock.take().expect("new_lock starts Some and can take at most once; qed");
      break;
    }
  }
}).expect("length of locks is not changed; qed");
if let Some(new_lock) = new_lock {
  locks.try_push(new_lock).map_err(|_| Error::<T, I>::TooManyLocks)?;
}
```